### DISCUSSION TITLE
Let's discuss React in Deno

### DISCUSSION THREAD

**Discussion by steida** [denoland/deno#13523] [Ideas]
I decided to start this discussion because I didn't find any official recommendations.

1) For example, how to switch dev and production builds?

2) How to write a React component library in Deno?

3) Without peerDependencies, how to ensure only one instance of React in an app? 

As for 3), one idea I have is to pass the whole React object to component factories. 
Such a component library would import only React namespace type. What do you think?

As I see it, the only alternative to peerDependendies is "types as a contract".

**Comment by steida** [comment]
```ts
// lib/deps.ts
export type { default as React } from "https://esm.sh/react@17.0.2";
```

```ts
// lib/ssr.tsx
import type { React } from "./deps.ts";

export const ssr =
  (deps: {
    renderToStaticMarkup: (element: React.ReactElement) => string;
    renderToString: (element: React.ReactElement) => string;
  }) =>
  (app: React.ReactElement): string => {
    return "todo";
  };
```

```ts
// in an app
ctx.response.body = ssr(ReactDOMServer)(<App />);
```

**Comment by steida** [comment]
Btw, should we use this or not? Does it have any advantage? 

```
{
  "imports": {
    "react": "https://esm.sh/react@17.0.2",
    "react-dom": "https://esm.sh/react-dom@17.0.2"
  }
}
```

**Comment by bradenmacdonald** [comment]
You can check out https://alephjs.org/ , they're building a whole framework for React in Deno just like Next.js (which is really awesome).

**Reply by steida** [reply]
Thank you, but I checked everything I was able to google. Aleph.js development has been stopped.
